### DISCUSSION TITLE
Provide a documentation for unit testing rules

### DISCUSSION THREAD

**Discussion by shouichi** [prometheus-operator/prometheus-operator#4233] [Ideas]
I want to run unit tests for alerting rules using `promtool`. But rules are defined in `PrometheusRule` CRD and it's not obvious to extract rules from CRD and run tests against them. A user guide explaining unit tests would be nice. Thanks!

References

- https://prometheus.io/docs/prometheus/latest/configuration/unit_testing_rules/

**Comment by paulfantom** [comment]
You can extract rules from CR with:
```
cat RULEFILE.yaml | gojsontoyaml -yamltojson | jq '.spec' | gojsontoyaml
```
Or if you already have PrometheusRule object in json (for example getting form k8s):
```
kubectl get prometheusrule MYRULES -o json | jq '.spec'
```

And proceed with testing via promtool.

`gojsontoyaml` binary is available from https://github.com/brancz/gojsontoyaml

**Reply by shouichi** [reply]
Thanks for the tip!

**Reply by kromanov-miro** [reply]
No need to do back-and-forth with jsons for yaml. [yq](https://github.com/mikefarah/yq) works great
`yq .spec < rules.yaml`
